pkg/common/consumer/consumer_manager.go

```diff
@@ -246,64 +278,123 @@ func (m *kafkaConsumerGroupManagerImpl) consume(ctx context.Context, groupId str
 
 // stopConsumerGroups closes the managed ConsumerGroup identified by the provided groupId, and marks it
 // as "stopped" (that is, "able to be restarted" as opposed to being closed by something outside the manager)
-func (m *kafkaConsumerGroupManagerImpl) stopConsumerGroup(groupId string) error {
+func (m *kafkaConsumerGroupManagerImpl) stopConsumerGroup(lock *commands.CommandLock, groupId string) error {
+	if err := m.processLock(lock, groupId, true); err != nil {
+		return err
+	}
+
 	groupLogger := m.logger.With(zap.String("GroupId", groupId))
 	groupLogger.Info("Stopping Managed ConsumerGroup")
-	managedGrp, ok := m.groups[groupId]
-	if ok {
-		// The managedGroup's start channel must be created (that is, the managedGroup must be marked
-		// as "stopped") before closing the internal ConsumerGroup, otherwise the consume function would
-		// return control to the factory.
-		managedGrp.createRestartChannel()
-		// Close the inner sarama ConsumerGroup, which will cause our consume() function to stop
-		// and wait for the managedGroup to start again.
-		err := managedGrp.group.Close()
-		if err != nil {
-			groupLogger.Error("Failed To Close Managed ConsumerGroup", zap.Error(err))
-			// Don't leave the start channel open if the group.Close() call failed; that would be misleading
-			managedGrp.closeRestartChannel()
-			return err
-		}
-		return nil
+
+	managedGrp := m.getGroup(groupId)
+	if managedGrp == nil {
+		groupLogger.Info("ConsumerGroup Not Managed - Ignoring Stop Request")
+		return fmt.Errorf("stop requested for consumer group not in managed list: %s", groupId)
+	}
+
+	// The managedGroup's start channel must be created (that is, the managedGroup must be marked
+	// as "stopped") before closing the internal ConsumerGroup, otherwise the consume function would
+	// return control to the factory.
+	managedGrp.createRestartChannel()
+	// Close the inner sarama ConsumerGroup, which will cause our consume() function to stop
+	// and wait for the managedGroup to start again.
+	err := managedGrp.group.Close()
+	if err != nil {
+		groupLogger.Error("Failed To Close Managed ConsumerGroup", zap.Error(err))
+		// Don't leave the start channel open if the group.Close() call failed; that would be misleading
+		managedGrp.closeRestartChannel()
+		return err
 	}
-	groupLogger.Info("ConsumerGroup Not Managed - Ignoring Stop Request")
-	return fmt.Errorf("stop requested for consumer group not in managed list: %s", groupId)
+
+	return m.processLock(lock, groupId, false)
```

Comment:
Comment here and above maybe?  not clear (yet - still reading) what processLock() does and how the initial boolean flag affects it?

Reply:
I'll add some hopefully demystifying comments.  The idea was that "processLock" would either lock or unlock the group based on the third argument ("before").  I originally had a "lockBefore" and "unlockAfter" function but they were 95% the same.  I am not opposed to recreating those and having them just call a common function if "processLock" is confusing

Reply:
I added back the lockBefore/unlockAfter wrapper functions; they're trivial but I think it makes reading the stopConsumerGroup and startConsumerGroup functions less confusing.